### DISCUSSION TITLE
change/improve the meaning of Awarded.HardcoreMode field on the database

### DISCUSSION THREAD

**Discussion by meleu** [RetroAchievements/RAWeb#724] [Ideas]
According to the database diagram @luchaos posted on #raweb the `Awarded` table has a field named `HardcoreMode` wich is a tinyint (then I guess 8 bits long).

Currently it seems to be used only as a boolean: `0` is "not-hardcore" (aka softcore) and `1` is hardcore.

I would like to propose another value-meaning: `2` - Unofficial.

(Question: possibly rename this field to simply `mode` in V2?)

### Usecase

Recently we changed the behavior of Unofficial achievements (PR #141) making them not be awardable on the site (therefore giving no points to the user). The main reason for this decision is to not give points when an achievement is not yet in the Core. But I recognize that having a timestamp of when a player earned the achievement can help at the development (and debugging) stage.

My intention is to use the "Unofficial mode" to register the timestamp of when an user got an achievement and **DO NOT** give the points.

Also, I want to make it be always "earnable", which means that when the user triggers the cheevo again the timestamp is updated (similar to what happens when users improve their score on a leaderboard). It would **NOT** return that error saying "User already has this achievement awarded."

**EDIT**
Here's the diagram:

![db_er_diagram-2](https://user-images.githubusercontent.com/8508804/41165384-dfa1ac78-6b13-11e8-87e0-33da37a667ee.png)

**Comment by meleu** [comment]
Another thought I had: use the `awarded.mode` field as bit flags.

- bit0: hardcore flag (which wouldn't break backward compatibility)
- bit1: unofficial flag
- bit2: manual award flag

A feature I currently miss is a way to see if a user got an achievement via manual unlock.

Sometimes we (admins) manually award achievements for users and it can confuse the devs when trying to solve their tickets (see #120).

Also, we can make it possible to allow the user to earn the achievement again normally by playing if the achievement is flagged as "manually awarded. It would just remove (set to zero) the manual award flag, update the timestamp and not award points (also not return an error).

**Comment by Jamiras** [comment]
None of the clients even try to call the server to unlock Unofficial achievements any more. If we wanted to track Unofficial unlocks, the clients would have to be updated.

I'm not sure that there's any merit in tracking Unofficial unlocks separately. If an unofficial unlock is requested, we can still put a normal unlock message in the Awarded table without updating the user's points. And when the points are recalculated, we'd just ignore achievements that are Unofficial (I think we do that already).

However, I agree that expanding the field to track manually awarded achievements makes sense.

**Reply by meleu** [reply]
Currently I don't think the implementation I'm suggesting here is a good idea.

IMO it makes more sense to track this kind of thing in the Awarded table but in a new field called, let's say, `client`. It could store the user-agents with emulators' version or manual-unlock info.

**Reply by Jamiras** [reply]
> However, I agree that expanding the field to track manually awarded achievements makes sense.

We do this now, but there isn't any way to view that information.

**Reply by Jamiras** [reply]
Manually awarded achievements are now tracked in the database, along with client User-Agents.
